lib/rucio/core/credential.py

```diff
@@ -107,4 +116,43 @@ def get_signed_url(service, operation, url, lifetime=600):
                                                                                                      lifetime,
                                                                                                      signature)
 
+    elif service == 's3':
+        # S3 can't sign delete URLs
+        if operation == 'delete':
+            raise UnsupportedOperation('Delete operation not supported for S3 signed URLs')
+
+        # split URL to get hostname, bucket and key
+        components = url.split('/')
+        if len(components) < 5:
+            raise UnsupportedOperation('Not a valid S3 URL')
+        host = components[2]
+        bucket = components[3]
+        key = '/'.join(components[4:])
+
+        # remove port number from host if present
+        colon = host.find(':')
+        if colon >= 0:
+            host = host[:colon]
+
+        # look up in RSE account configuration by <hostname>_<bucketname>
+        cred_name = host + "_" + bucket
+        cred = REGION.get('s3-%s' % cred_name)
+        if cred is NO_VALUE:
+            rse_cred = get_rse_credentials()
+            cred = rse_cred.get(cred_name)
+            REGION.set('s3-%s' % cred_name, cred)
+        access_key = cred['access_key']
+        secret_key = cred['secret_key']
+        signature_version = cred['signature_version']
+        region_name = cred['region']
+
+        if operation == 'read':
+            s3op = 'get_object'
+        else:
+            s3op = 'put_object'
+
+        s3 = boto3.client('s3', aws_access_key_id=access_key, aws_secret_access_key=secret_key, config=Config(signature_version=signature_version, region_name=region_name))
```

Comment:
can you wrap the important bits in a `core.monitor.record_timer_block` so we can see the number of signatures (and how long it takes to sign) in graphite?

Reply:
I have added a `record_timer_block` around the signing.